<h1>QArithmetic.py</h1>
A Qiskit library for binary arithmetic and bitwise operations.

<ul>
    <li>Mohamed Hassan, Virginia Tech</li>
    <li>Harsh Khetawat, North Carolina State University</li>
    <li>Ana Neri, University of Minho</li>
    <li>Afonso Rodrigues, University of Minho</li>
    <li>Tom Wong, Creighton University</li>
</ul>

<h2>Table of Contents</h2>
<ol>
    <li>Motivation</li>
    <li>Binary Arithmetic</li>
    <ol>
        <li>Add</li>
        <li>Subtract</li>
        <li>Multiply</li>
        <li>Divide and Remainder</li>
    </ol>
    <li>Bitwise Operations</li>
    <ol>
        <li>AND</li>
        <li>OR</li>
        <li>XOR</li>
        <li>NOT</li>
    </ol>
    <li>Other Functions</li>
</ol>

<h1> Motivation </h1>

<ol>
    <li>Creating Oracles</li>
    <li>Finance Applications</li>
</ol>

<h1>Binary Arithmetic</h1>
<h2>Add</h2>
|a⟩|b⟩ &rarr; |a⟩|a+b⟩

In [5]:
from qiskit import QuantumCircuit, ClassicalRegister, QuantumRegister
from qiskit import execute, Aer
from QArithmetic import add

# Registers and circuit.
a = QuantumRegister(5)
b = QuantumRegister(5)
ca = ClassicalRegister(5)
cb = ClassicalRegister(5)
qc = QuantumCircuit(a, b, ca, cb)

# Numbers to add.
qc.x(a[1]) # a = 01110
qc.x(a[2])
qc.x(a[3])
qc.x(b[0]) # b = 01011
qc.x(b[1])
qc.x(b[3])

# Add the numbers, so |a>|b> to |a>|a+b>.
add(qc, a, b, 5)

# Measure the results.
qc.measure(a, ca)
qc.measure(b, cb)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'11001 01110': 1024}


Thus, 01110 + 01011 = 11001. That is, 14 + 11 = 25, as expected.

<h2>Subtract</h2>
|a⟩|b⟩ &rarr; |a⟩|a-b⟩

In [6]:
from QArithmetic import sub

# Registers and circuit.
a = QuantumRegister(5)
b = QuantumRegister(5)
ca = ClassicalRegister(5)
cb = ClassicalRegister(5)
qc = QuantumCircuit(a, b, ca, cb)

# Numbers to subtract.
qc.x(a[1]) # a = 01110
qc.x(a[2])
qc.x(a[3])
qc.x(b[0]) # b = 01011
qc.x(b[1])
qc.x(b[3])

# Add the numbers, so |a>|b> to |a>|a-b>.
sub(qc, a, b, 5)

# Measure the results.
qc.measure(a, ca)
qc.measure(b, cb)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'00011 01110': 1024}


Thus, 01110 - 01011 = 00011. That is, 14 - 11 = 3, as expected.

<h2>Multiply</h2>
|a⟩|b⟩|0...0> &rarr; |a⟩|b⟩|a*b⟩

In [7]:
from QArithmetic import mult

# Registers and circuit.
a = QuantumRegister(3)
b = QuantumRegister(3)
m = QuantumRegister(6)
cm = ClassicalRegister(6)
qc = QuantumCircuit(a, b, m, cm)

# Numbers to multiply.
qc.x(a[1]) # a = 010 = 2
qc.x(b[0]) # b = 011 = 3
qc.x(b[1])

# Multiply the numbers, so |a>|b>|m=0> to |a>|b>|a*b>.
mult(qc, a, b, m, 3)

# Measure the result.
qc.measure(m, cm)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))


{'000110': 1024}


Thus, 010 * 011 = 000110. That is, 2 x 3 = 6, as expected.

<h2>Divide and Remainder</h2>
|0...0 a⟩|b 0...0⟩|0...0⟩ &rarr; |a%b 0...0⟩|b⟩|a/b⟩

In [21]:
from QArithmetic import div

# Registers and circuit.
a = QuantumRegister(4)
b = QuantumRegister(4)
c = QuantumRegister(2)
ca = ClassicalRegister(4)
cb = ClassicalRegister(4)
cc = ClassicalRegister(2)
qc = QuantumCircuit(a,b,c,ca,cb,cc)

# Inputs.
qc.x(a[0]) # a = 0011
qc.x(a[1])
qc.x(b[3]) # b = 1000

# Divide a and b.
div(qc, a, b, c, 2)

# Measure.
qc.measure(a, ca)
qc.measure(b, cb)
qc.measure(c, cc)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'01 1000 0100': 1024}


Thus, 11 / 10 = 01 remainder 01. That is, 3/2 = 1 remainder 1, as expected.

<h1>Bitwise Operations</h1>
<h2>AND</h2>
|a⟩|b⟩|0...0> &rarr; |a⟩|b⟩|a and b⟩

In [14]:
from QArithmetic import bitwise_and

# Registers and circuit.
a = QuantumRegister(4)
b = QuantumRegister(4)
c = QuantumRegister(4)
ca = ClassicalRegister(4)
cb = ClassicalRegister(4)
cc = ClassicalRegister(4)
qc = QuantumCircuit(a, b, c, ca, cb, cc)

# Inputs
qc.x(a[1]) # a =1010
qc.x(a[3])
qc.x(b[0]) # b = 1011
qc.x(b[1])
qc.x(b[3])

# Take the bitwise AND.
bitwise_and(qc, a, b, c, 4)

# Measure.
qc.measure(a, ca)
qc.measure(b, cb)
qc.measure(c, cc)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'1010 1011 1010': 1024}


Thus, 1010 & 1011 = 1010, as expected.

<h2>OR</h2>
|a⟩|b⟩|0...0> &rarr; |a⟩|b⟩|a or b⟩

In [15]:
from QArithmetic import bitwise_or

# Registers and circuit.
a = QuantumRegister(4)
b = QuantumRegister(4)
c = QuantumRegister(4)
ca = ClassicalRegister(4)
cb = ClassicalRegister(4)
cc = ClassicalRegister(4)
qc = QuantumCircuit(a, b, c, ca, cb, cc)

# Input.
qc.x(a[1]) # a = 1010
qc.x(a[3])
qc.x(b[0]) # b = 1011
qc.x(b[1])
qc.x(b[3])

# Take the bitwise OR.
bitwise_or(qc, a, b, c, N)

# Measure.
qc.measure(a, ca)
qc.measure(b, cb)
qc.measure(c, cc)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'1011 1011 1010': 1024}


Thus, 1010 & 1011 = 1011, as expected.

<h2>XOR</h2>
|a⟩|b⟩|0...0> &rarr; |a⟩|b⟩|a xor b⟩

In [16]:
from QArithmetic import bitwise_xor

# Registers and circuit.
a = QuantumRegister(4)
b = QuantumRegister(4)
c = QuantumRegister(4)
ca = ClassicalRegister(4)
cb = ClassicalRegister(4)
cc = ClassicalRegister(4)
qc = QuantumCircuit(a, b, c, ca, cb, cc)

# Input.
qc.x(a[1]) # a = 1010
qc.x(a[3])
qc.x(b[0]) # b = 1011
qc.x(b[1])
qc.x(b[3])

# Take the bitwise XOR.
bitwise_xor(qc, a, b, c, N)

# Measure.
qc.measure(a, ca)
qc.measure(b, cb)
qc.measure(c, cc)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'0001 1011 1010': 1024}


Thus, 1010 & 1011 = 1011, as expected.

<h2>NOT</h2>
|a⟩|0...0> &rarr; |a⟩|~a⟩

In [17]:
from QArithmetic import bitwise_not

# Registers and circuit.
a = QuantumRegister(4)
b = QuantumRegister(4)
ca = ClassicalRegister(4)
cb = ClassicalRegister(4)
qc = QuantumCircuit(a, b, ca, cb)

# Input.
qc.x(a[1]) # a =  1010
qc.x(a[3])

# Take the bitwise NOT.
bitwise_not(qc, a, b, N)

# Measure.
qc.measure(a, ca)
qc.measure(b, cb)

# Simulate the circuit.
backend_sim = Aer.get_backend('qasm_simulator')
job_sim = execute(qc, backend_sim)
result_sim = job_sim.result()

print(result_sim.get_counts(qc))

{'0101 1010': 1024}


Thus, ~1010 = 0101, as expected.

<h2>Other Functions</h2>
<ol>
    <li>Bitwise Cyclic Shifts</li>
    <li>Controlled-Add</li>
    <li>Controlled-Controlled-Controlled-NOT (or Controlled Toffoli)</li>
    <li>Quantum Fourier transform and inverse</li>
    <li>Controlled quantum Fourier transform and inverse</li>
</ol>

<h2>Summary</h2>
<ol>
    <li>Motivation</li>
    <li>Binary Arithmetic</li>
    <ol>
        <li>Add</li>
        <li>Subtract</li>
        <li>Multiply</li>
        <li>Divide and Remainder</li>
    </ol>
    <li>Bitwise Operations</li>
    <ol>
        <li>AND</li>
        <li>OR</li>
        <li>XOR</li>
        <li>NOT</li>
    </ol>
    <li>Other Functions</li>
</ol>

Thank you!